src/kernel/arch/x86/pit.zig

```diff
@@ -0,0 +1,288 @@
+///
+/// Arguments:
+///     IN cmd: u8 - The command to send to the PIT.
+///
+inline fn sendCommand(cmd: u8) void {
+    arch.outb(COMMAND_REGISTER, cmd);
+}
+
+///
+/// Send data to a given counter. Will be only one of the 3 counters as is an internal function.
+///
+/// Arguments:
+///     IN comptime counter: u16 - The counter port to send the data to.
+///     IN data: u8              - The data to send.
+///
+inline fn sendDateToCounter(comptime counter: u16, data: u8) void {
+    assert(counter == COUNTER_0_REGISTER or counter == COUNTER_1_REGISTER or counter == COUNTER_2_REGISTER);
+    arch.outb(counter, data);
+}
+
+fn pitHandler(context: *arch.InterruptContext) void {
+    ticks += 1;
+}
+
+///
+/// Set up a counter with a tick rate and mode of operation.
+///
+/// Arguments:
+///     IN freq: u16   - The frequency that the counter operates at.
+///     IN counter: u8 - Which counter is to be set up, either OCW_SELECT_COUNTER_0,
+///                      OCW_SELECT_COUNTER_1 or OCW_SELECT_COUNTER_2 only.
+///     IN mode: u8    - The mode of operation that the counter will operate in. See The modes
+///                      definition above to chose which mode the counter is to run at.
+///
+fn setupCounter(comptime counter: u8, freq: u32, mode: u8) void {
+    var reload_value: u32 = 0x10000; // 65536, the slowest possible frequency
+
+    if (freq > 18) {
+        // The fastest possible frequency if frequency is too high
+        reload_value = 1;
+
+        if (freq < MAX_FREQUENCY) {
+            reload_value = MAX_FREQUENCY / freq;
+            var rem: u32 = MAX_FREQUENCY % freq;
+
+            // Is the remainder more than half
+            if (rem > MAX_FREQUENCY / 2) {
+                // Round up
+                reload_value += 1;
+            }
+        }
+    }
+
+    // Get the u16 version as this is what will be loaded into the PIT
+    var reload_val_16: u16 = @truncate(u16, reload_value);
+
+    // Update the frequency with the actual one that the PIT will be using
+    var frequency: u32 = MAX_FREQUENCY / reload_value;
+    var rem: u32 = MAX_FREQUENCY % reload_value;
+
+    // Is the remainder more than half
+    if (rem > MAX_FREQUENCY / 2) {
+        // Round up
+        frequency += 1;
+    }
+
+    // Calculate the amount of time between IRQs
+    time_ms = reload_value * u32(1000);
+    time_under_1_ms = time_ms % MAX_FREQUENCY;
+    time_ms /= MAX_FREQUENCY;
+
+    comptime const port: u16 = switch (counter) {
+        OCW_SELECT_COUNTER_0 => COUNTER_0_REGISTER,
+        OCW_SELECT_COUNTER_1 => COUNTER_1_REGISTER,
+        OCW_SELECT_COUNTER_2 => COUNTER_2_REGISTER,
+        else => unreachable,
+    };
+
+    switch (counter) {
+        OCW_SELECT_COUNTER_0 => current_freq_0 = frequency,
+        OCW_SELECT_COUNTER_1 => current_freq_1 = frequency,
+        OCW_SELECT_COUNTER_2 => current_freq_2 = frequency,
+        else => unreachable,
+    }
+
+    var command: u8 = 0;
```

Comment:
Not a big deal but you could assign it to `mode` instead of assigning it to 0 then or'ing with `mode`.

Reply:
Can do, less code